### PR TITLE
RSDK-5799

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-5799)

- Fix issue where `kinematicBase.Stop()` would be called with a cancelled context, which becomes a no op
- Ensure `kinematicBase.Stop()` is always called with a fresh context if context is canceled in move request or if GoToInputs returns an error

Tested using gazebo simulation:
https://github.com/viamrobotics/rdk/assets/5927876/78dc8aba-fe82-4024-ae1b-653b769fae58

and the following script:

```go
package main

import (
	"context"
	"encoding/json"
	"math"
	"os"
	"os/signal"
	"strconv"
	"syscall"

	"github.com/edaniels/golog"
	geo "github.com/kellydunn/golang-geo"
	"go.viam.com/rdk/components/base"
	"go.viam.com/rdk/components/movementsensor"
	"go.viam.com/rdk/robot/client"
	"go.viam.com/rdk/services/motion"
	"go.viam.com/utils/rpc"
)

type s struct {
	dest *geo.Point
}

func main() {
	shutdown := make(chan os.Signal, 1)
	done := make(chan struct{}, 1)
	signal.Notify(shutdown, syscall.SIGINT, syscall.SIGTERM)
	logger := golog.NewDevelopmentLogger("client")
	if len(os.Args) != 3 {
		logger.Error("usage: <cmd> lat lng")
		return
	}
	lat, err := strconv.ParseFloat(os.Args[1], 64)
	if err != nil {
		logger.Errorf("lat not able to be parsed %s", err)
	}

	lng, err := strconv.ParseFloat(os.Args[2], 64)
	if err != nil {
		logger.Errorf("lng not able to be parsed %s", err)
	}

	s := s{dest: geo.NewPoint(lat, lng)}

	robot, err := client.New(
		context.Background(),
		"",
		logger,
		client.WithDialOptions(rpc.WithEntityCredentials(
			"",
			rpc.Credentials{
				Type:    rpc.CredentialsTypeAPIKey,
				Payload: "",
			})),
	)
	if err != nil {
		logger.Fatal(err)
	}

	defer robot.Close(context.Background())
	// logger.Info("Resources:")
	// logger.Info(robot.ResourceNames())

	merged, err := movementsensor.FromRobot(robot, "merged")
	if err != nil {
		logger.Error(err)
		return
	}
	m, err := motion.FromRobot(robot, "builtin")
	if err != nil {
		logger.Error(err)
		return
	}
	position, _, err := merged.Position(context.Background(), nil)
	if err != nil {
		logger.Error(err)
		return
	}

	heading, err := merged.CompassHeading(context.Background(), nil)
	if err != nil {
		logger.Error(err)
		return
	}

	b, err := json.Marshal(map[string]interface{}{"type": "start", "heading": heading, "position": map[string]float64{"lat": position.Lat(), "lng": position.Lng()}})
	if err != nil {
		logger.Error(err)
		return
	}
	logger.Info(string(b))

	b, err = json.Marshal(map[string]interface{}{"type": "goal", "position": map[string]float64{"lat": s.dest.Lat(), "lng": s.dest.Lng()}})
	if err != nil {
		logger.Error(err)
		return
	}
	logger.Info(string(b))
	ctx, cancelFn := context.WithCancel(context.Background())
	go func() {
		select {
		case <-shutdown:
		case <-done:
		}
		cancelFn()
		logger.Info("shutting down")
	}()

	res, err := m.MoveOnGlobe(
		ctx,
		base.Named("base"),
		s.dest,
		math.NaN(),
		movementsensor.Named("merged"),
		nil,
		&motion.MotionConfiguration{
			PositionPollingFreqHz: 1,
			ObstaclePollingFreqHz: 2,
			PlanDeviationMM:       26000,
			LinearMPerSec:         0.3,
			AngularDegsPerSec:     20,
		},
		map[string]interface{}{"smooth_iter": 20},
	)
	if err != nil {
		logger.Error(err)
		return
	}
	logger.Infof("res: %v", res)
	done <- struct{}{}
}
```

### CODE REVIEW

File: services/motion/builtin/move_request.go [hunk 1]
```diff
@@ -124,16 +125,27 @@ func (mr *moveRequest) execute(ctx context.Context, waypoints state.Waypoints, w
 	for i := int(waypointIndex.Load()); i < len(waypoints); i++ {
 		select {
 		case <-ctx.Done():
+			mr.logger.Debugf("calling kinematicBase.Stop due to %s\n", ctx.Err())
+			stopCtx, cancelFn := context.WithTimeout(context.Background(), baseStopTimeout)
+			if stopErr := mr.kinematicBase.Stop(stopCtx, nil); stopErr != nil {
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:130]: I think we can defer the cancelFn here instead

File: services/motion/builtin/move_request.go [hunk 2]
```diff
@@ -124,16 +125,27 @@ func (mr *moveRequest) execute(ctx context.Context, waypoints state.Waypoints, w
...
+			cancelFn()
 			return state.ExecuteResponse{}, nil
 		default:
 			mr.planRequest.Logger.Info(waypoints[i])
 			if err := mr.kinematicBase.GoToInputs(ctx, waypoints[i]); err != nil {
 				// If there is an error on GoToInputs, stop the component if possible before returning the error
 				mr.logger.Debugf("calling kinematicBase.Stop due to %s\n", err)
-				if stopErr := mr.kinematicBase.Stop(ctx, nil); stopErr != nil {
+				stopCtx, cancelFn := context.WithTimeout(context.Background(), baseStopTimeout)
+				if stopErr := mr.kinematicBase.Stop(stopCtx, nil); stopErr != nil {
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:143]: Looks like we're reusing the same code block here.  I think making a function like `motionRequest.Stop` could be a good improvement

**Review by raybjork — CHANGES_REQUESTED** [review]
Just a couple nits

File: services/motion/builtin/move_request.go (hunk 1 above)
Comment (nicksanford) [reply services/motion/builtin/move_request.go:130]: the linter won't let me, I'll paste the error in a moment

Comment (nicksanford) [reply services/motion/builtin/move_request.go]: done

File: services/motion/builtin/move_request.go (hunk 2 above)
Comment (nicksanford) [reply services/motion/builtin/move_request.go]: done

File: services/motion/builtin/move_request.go [hunk 3]
```diff
@@ -124,14 +125,17 @@ func (mr *moveRequest) execute(ctx context.Context, waypoints state.Waypoints, w
...
+			mr.logger.Debugf("calling kinematicBase.Stop due to %s\n", ctx.Err())
+			if stopErr := mr.stop(); stopErr != nil {
+				return state.ExecuteResponse{}, errors.Wrap(ctx.Err(), stopErr.Error())
+			}
 			return state.ExecuteResponse{}, nil
 		default:
 			mr.planRequest.Logger.Info(waypoints[i])
 			if err := mr.kinematicBase.GoToInputs(ctx, waypoints[i]); err != nil {
 				// If there is an error on GoToInputs, stop the component if possible before returning the error
 				mr.logger.Debugf("calling kinematicBase.Stop due to %s\n", err)
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:137]: I would also bring this inside the function since it exists in two places still

**Review by raybjork — APPROVED** [review]
I just have one more nit but after that I'm good with merging

**Pull request merged**